worker/caasoperator/action.go

```diff
@@ -38,48 +42,120 @@ func ensurePath(
 	return errors.Trace(err)
 }
 
+func ensureSymlink(
+	client exec.Executor,
+	podName string,
+	oldName, newName string,
+	stdout io.Writer,
+	stderr io.Writer,
+	cancel <-chan struct{},
+) error {
+	logger.Debugf("making symlink %v->%v", newName, oldName)
+	err := client.Exec(
+		exec.ExecParams{
+			PodName:  podName,
+			Commands: []string{"test", "-f", newName, "||", "ln", "-s", oldName, newName},
+			Stdout:   stdout,
+			Stderr:   stderr,
+		},
+		cancel,
+	)
+	return errors.Trace(err)
+}
+
+type unitOperatorInfo struct {
+	OperatorAddress string `yaml:"operator-address"`
+}
+
 func prepare(
 	client exec.Executor,
 	podName string,
+	serviceAddress string,
 	operatorPaths Paths,
 	unitPaths uniter.Paths,
 	stdout io.Writer,
 	stderr io.Writer,
 	cancel <-chan struct{},
 ) error {
-	for _, path := range []string{
-		// order matters here.
-		operatorPaths.GetCharmDir(),
-		unitPaths.GetCharmDir(),
-
-		filepath.Join(operatorPaths.GetToolsDir(), "jujud"),
-		unitPaths.GetToolsDir(),
+	type workloadPathSpec struct {
+		src, dest string
+	}
+	// Copy the core charm files and jujud binary.
+	for _, pathSpec := range []workloadPathSpec{
+		{
+			src:  operatorPaths.GetCharmDir(),
+			dest: unitPaths.State.BaseDir,
+		}, {
+			src:  filepath.Join(operatorPaths.GetToolsDir(), "jujud"),
+			dest: unitPaths.ToolsDir,
+		},
 	} {
 
-		_, err := os.Stat(path)
+		_, err := os.Stat(pathSpec.src)
 		if os.IsNotExist(err) {
-			return errors.NotFoundf("file or path %q", path)
+			return errors.NotFoundf("file or path %q", pathSpec.src)
 		}
 		if err != nil {
 			return errors.Trace(err)
 		}
-		destPath := filepath.Dir(path)
-		if err := ensurePath(client, podName, destPath, stdout, stderr, cancel); err != nil {
+		logger.Debugf("copy path %q to %q", pathSpec.src, pathSpec.dest)
+		if err := ensurePath(client, podName, pathSpec.dest, stdout, stderr, cancel); err != nil {
 			return errors.Trace(err)
 		}
 
 		if err := client.Copy(exec.CopyParam{
 			Src: exec.FileResource{
-				Path: path,
+				Path: pathSpec.src,
 			},
 			Dest: exec.FileResource{
-				Path:    destPath,
+				Path:    pathSpec.dest,
 				PodName: podName,
 			},
 		}, cancel); err != nil {
 			return errors.Trace(err)
 		}
 	}
+
+	// Create the operator.yaml file containing the operator service address.
+	if err := ensurePath(client, podName, unitPaths.State.BaseDir, stdout, stderr, cancel); err != nil {
+		return errors.Trace(err)
+	}
+	opInfo := unitOperatorInfo{OperatorAddress: serviceAddress}
```

Comment:
write the operatorInfo file into unit dir(charm dir or somewhere), so it will be synced from above line at [] workloadPathSpec?

Reply:
we don't want to pollute the files laid out already in the operator pod